src/client/Client.cc

```diff
@@ -2077,37 +2077,66 @@ void Client::handle_client_reply(MClientReply *reply)
     mount_cond.Signal();
 }
 
+void Client::_handle_full_flag(int64_t pool)
+{
+  ldout(cct, 1) << __func__ << ": FULL: cancelling outstanding operations "
+    << "on " << pool << dendl;
+  // Cancel all outstanding ops with -ENOSPC: it is necessary to do this rather
+  // than blocking, because otherwise when we fill up we potentially lock caps
+  // forever on files with dirty pages, and we need to be able to release
+  // those caps to the MDS so that it can delete files and free up space.
+  epoch_t cancelled_epoch = objecter->op_cancel_writes(-ENOSPC, pool);
+
+  // For all inodes with a pending flush write op (i.e. one of the ones we
+  // will cancel), we've got to purge_set their data from ObjectCacher
+  // so that it doesn't re-issue the write in response to the ENOSPC error.
+  // Fortunately since we're cancelling *everything*, we don't need to know
+  // which ops belong to which ObjectSet, we can just blow all the un-flushed
```

Comment:
This comment is no longer accurate; we might not be canceling everything. I think the code is handling it even so but I'm not certain if I'm missing something about the op cancellation elsewhere.